File: problems/235. Lowest Common Ancestor of a Binary Search Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    lowerBound = None
    upperBound = None

    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        self.lowerBound, self.upperBound = min(p.val, q.val), max(p.val, q.val)
        height = self.getHeight(root)
        lca = None
        for i in range(1, height + 1):
            lca = self.levelOrder(i, root)
            if lca != None: return lca
        return None

    def getHeight(self, root: 'TreeNode') -> int:
        if not root: return 0
        leftHeight = 0 if root.left == None else self.getHeight(root.left)
        rightHeight = 0 if root.right == None else self.getHeight(root.right)
        return max(leftHeight, rightHeight) + 1

    def levelOrder(self, height: int, root: 'TreeNode') -> Optional[TreeNode]:
        if not root: return None
        if (height == 1) and (root.val >= self.lowerBound) and (root.val <= self.upperBound):
            return root
        else:
            leftRet = self.levelOrder(height - 1, root.left)
            rightRet = self.levelOrder(height - 1, root.right)
            if leftRet or rightRet: return leftRet if leftRet else rightRet
            else: return None
        

File: problems/235. Lowest Common Ancestor of a Binary Search Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    private int lowerBound = 0;
    private int upperBound = 0;

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        lowerBound = Math.min(p.val, q.val);
        upperBound = Math.max(p.val, q.val);
        int height = getHeight(root);
        for (int i = 1; i <= height; i++) {
            TreeNode lca = levelOrder(i, root);
            if (lca != null) return lca;
        }
        return null;
    }

    private int getHeight(TreeNode root) {
        if (root == null) return 0;
        int leftHeight = (root.left == null) ? 0 : getHeight(root.left);
        int rightHeight = (root.right == null) ? 0 : getHeight(root.right);
        return Math.max(leftHeight, rightHeight) + 1;
    }

    private TreeNode levelOrder(int height, TreeNode root) {
        if (root == null) return null;
        if (height == 1 && root.val >= lowerBound && root.val <= upperBound) {
            return root;
        } else {
            TreeNode left = levelOrder(height - 1, root.left);
            TreeNode right = levelOrder(height - 1, root.right);
            if (left != null || right != null) return left == null ? right : left;
            else return null;
        }
    }
}